code/extensions/FolderSecured.php

```diff
@@ -1,201 +1,77 @@
-                // Note, we do this in the database rather than object-model; otherwise we get all sorts of problems
-                // about deleting files
-                $children = DB::query("SELECT \"ID\" FROM \"File\""
-                    . " WHERE \"Name\" = '$childName' AND \"ParentID\" = $parentID")->column();
-                if($children) {
-                    $keptChild = array_shift($children);
-                    foreach($children as $removedChild) {
-                        DB::query("UPDATE \"File\" SET \"ParentID\" = $keptChild WHERE \"ParentID\" = $removedChild");
-                        DB::query("DELETE FROM \"File\" WHERE \"ID\" = $removedChild");
-                    }
-                } else {
-                    user_error("Inconsistent database issue: SELECT ID FROM \"File\" WHERE Name = '$childName'"
-                        . " AND ParentID = $parentID should have returned data", E_USER_WARNING);
-                }
-            }
-        }
-
-        // Get index of database content
-        // We don't use DataObject so that things like subsites doesn't muck with this.
-        $dbChildren = DB::query("SELECT * FROM \"File\" WHERE \"ParentID\" = $parentID");
-        $hasDbChild = array();
-
-        if($dbChildren) {
-            foreach($dbChildren as $dbChild) {
-                $className = $dbChild['ClassName'];
-                if(!$className) $className = "File";
-                $hasDbChild[$dbChild['Name']] = new $className($dbChild);
-            }
+    public function augmentSQL(SQLQuery &$query) {
+        $sql = $query->sql();
+        $rawQuery = str_replace(array('"', "'"), '', $sql);
+        $isDuplicateSql = stristr($rawQuery, "SELECT DISTINCT File.ClassName,") !== false;
+        $isStdFilesSyncReq = $this->isStandardFileSyncRequest();
+        if($isStdFilesSyncReq && $isDuplicateSql) {
+            $query->addWhere('"Secured" = \'0\'');
         }
-        $unwantedDbChildren = $hasDbChild;
+    }
 
-        // if we're syncing a folder with no ID, we assume we're syncing the root assets folder
-        // however the Filename field is populated with "NewFolder", so we need to set this to empty
-        // to satisfy the baseDir variable below, which is the root folder to scan for new files in
-        if(!$parentID) {
-            $this->owner->Filename = '';
+    /**
+     * This method hooks into the write that is performed by {@link Folder::constructChild()} which is
+     * called by the main {@link Folder::syncChildren()} method. This simply ensures that the correct value for
+     * the "File" table's "Secured" field is written-to.
+     *
+     * @param array $manipulation
+     * @return null|void
+     */
+    public function augmentWrite(&$manipulation) {
+        // Don't bother with remaining logic if the controller request is not a synchronisation request
+        $isStdFilesSyncReq = $this->isStandardFileSyncRequest();
+        if(!$isStdFilesSyncReq) {
```

Comment:
since the function isStandardFileSyncRequest() checks isAssetsRoot, but this argumentWrite() should be applied to anywhere, not only when on "assets" folder. so we should move out isAssetsRoot check block out of isStardardFileSyncRequest() and probably make a new function isAssetsRoot(). and only check it in argmentSQL().